README.md
# `color_adjust` Flutter Extension Documentation

The `color_adjust` is an extension on the Flutter Color class that provides a collection of color manipulation methods.

## Usage

To use these methods, you need to import the `color_adjust` extension and apply these methods on a Color object.

![Screenshot of Example](https://user-images.githubusercontent.com/83240961/248578254-c5785210-fb46-4f1e-8ac8-fecaec034512.png)

Example:

```dart
import 'package:color_adjust/color_adjust.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    Color baseColor = Colors.blue;
    Color mixColor = Colors.pink;

    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text("Color Adjust Examples"),
        ),
        body: ListView(
          children: <Widget>[
            Container(color: baseColor, height: 50),
            Container(color: baseColor.darken(0.2), height: 50),
            Container(color: baseColor.lighten(0.2), height: 50),
            Container(color: baseColor.invert(), height: 50),
            Container(color: baseColor.compliment(), height: 50),
            Container(color: baseColor.desaturate(0.2), height: 50),
            Container(color: baseColor.saturate(0.2), height: 50),
            Container(color: baseColor.mix(mixColor, 0.5), height: 50),
            Container(color: baseColor.whiten(0.5), height: 50),
            Container(color: baseColor.tone(0.5), height: 50),
            Container(color: baseColor.blacken(0.5), height: 50),
          ],
        ),
      ),
    );
  }
}

```

## Methods

The extension adds the following methods:

- **`darken([double amount = .5])`**: Darkens the color by the provided amount (0-1), defaulting to 50%.
- **`lighten([double amount = .5])`**: Lightens the color by the provided amount (0-1), defaulting to 50%.
- **`invert()`**: Inverts a color, producing its "negative". It does not preserve lightness. If you want to do that, use the `compliment` method.
- **`compliment()`**: Gets the complimentary (opposite on the color wheel) of a color. This is like inverting, however it keeps the same lightness value.
- **`desaturate([double amount = .5])`**: Desaturates a color by the provided amount (0-1), defaulting to 50%.
- **`saturate([double amount = .5])`**: Saturates a color by the provided amount (0-1), defaulting to 50%.
- **`mix(Color another, [double amount = 0.5])`**: Mixes a color with another one by the provided amount (0-1), defaulting to 50%.
- **`whiten([double amount = 0.5])`**: Whitens a color by mixing it with white by the provided amount (0-1), defaulting to 50%.
- **`tone([double amount = 0.5])`**: Tones down a color by mixing it with grey by the provided amount (0-1), defaulting to 50%.
- **`blacken([double amount = 0.5])`**: Blackens a color by mixing it with black by the provided amount (0-1), defaulting to 50%.
- **`isDark`**: Checks if a color is considered dark using the luminance.
- **`isBright`**: Checks if a color is considered bright.


CHANGELOG.md
1.0.0 new
1.0.1 release
1.0.2 upadted readme
1.0.3 dependency fix
1.0.4 screenshot
1.0.5 fixed comments


lib/adjust.dart
import 'package:flutter/material.dart';

// Extension on Color to add color manipulation methods
extension Adjust on Color {
  // Darken a color by the provided amount (0-1), defaulting to 50%
  Color darken([double amount = .5]) {
    assert(amount >= 0 && amount <= 1);
    final hsl = HSLColor.fromColor(this); // convert to HSL
    // darken by reducing lightness
    final hslDark = hsl.withLightness((hsl.lightness - amount).clamp(0.0, 1.0));
    return hslDark.toColor(); // convert back to Color
  }

  // Lighten a color by the provided amount (0-1), defaulting to 50%
  Color lighten([double amount = .5]) {
    assert(amount >= 0 && amount <= 1);
    final hsl = HSLColor.fromColor(this); // convert to HSL
    // lighten by increasing lightness
    final hslLight =
        hsl.withLightness((hsl.lightness + amount).clamp(0.0, 1.0));
    return hslLight.toColor(); // convert back to Color
  }

  // Invert a color (producing its "negative")
  // This will not preserve lightness. If you want to do that, use compliment
  // to convert to HSL, invert the hue, and convert back to RGB.
  Color invert() {
    final r = (255 - red).clamp(0, 255);
    final g = (255 - green).clamp(0, 255);
    final b = (255 - blue).clamp(0, 255);
    return Color.fromARGB((opacity * 255).round(), r, g, b);
  }

  // Get the complimentary (opposite on the color wheel) of a color
  // This is like inverting, however it keeps the same lightness value
  Color compliment() {
    final hsl = HSLColor.fromColor(this); // convert to HSL
    final double newHue =
        (hsl.hue + 180.0) % 360.0; // compute complimentary hue
    final newHsl =
        hsl.withHue(newHue); // create new color with complimentary hue
    return newHsl.toColor(); // convert back to Color
  }

  // Desaturate a color by the provided amount (0-1), defaulting to 50%
  Color desaturate([double amount = .5]) {
    assert(amount >= 0 && amount <= 1);
    final hsl = HSLColor.fromColor(this); // convert to HSL
    // desaturate by reducing saturation
    final hslDesaturated =
        hsl.withSaturation((hsl.saturation - amount).clamp(0.0, 1.0));
    return hslDesaturated.toColor(); // convert back to Color
  }

  // Saturate a color by the provided amount (0-1), defaulting to 50%
  Color saturate([double amount = .5]) {
    assert(amount >= 0 && amount <= 1);
    final hsl = HSLColor.fromColor(this); // convert to HSL
    // saturate by increasing saturation
    final hslSaturated =
        hsl.withSaturation((hsl.saturation + amount).clamp(0.0, 1.0));
    return hslSaturated.toColor(); // convert back to Color
  }

  // Mix a color with another one by the provided amount (0-1), defaulting to 50%
  // this might result in a color that looks muddy, as the lightness is not preserved
  // if you want to preserve the lightness, use blend instead
  Color mix(Color another, [double amount = 0.5]) {
    assert(amount >= 0 && amount <= 1);
    // mix by taking a weighted average
    final r = (red * (1 - amount) + another.red * amount).round().clamp(0, 255);
    final g =
        (green * (1 - amount) + another.green * amount).round().clamp(0, 255);
    final b =
        (blue * (1 - amount) + another.blue * amount).round().clamp(0, 255);
    return Color.fromARGB(alpha, r, g, b);
  }

  // Mix a color with another one using HSL color space by the provided amount (0-1), defaulting to 50%
  Color blend(Color another, [double amount = 0.5]) {
    assert(amount >= 0 && amount <= 1);
    final hsl1 = HSLColor.fromColor(this);
    final hsl2 = HSLColor.fromColor(another);
    // mix by taking a weighted average
    final h =
        ((hsl1.hue * (1 - amount)) + (hsl2.hue * amount)).clamp(0.0, 360.0);
    final s = ((hsl1.saturation * (1 - amount)) + (hsl2.saturation * amount))
        .clamp(0.0, 1.0);
    final l = ((hsl1.lightness * (1 - amount)) + (hsl2.lightness * amount))
        .clamp(0.0, 1.0);
    return HSLColor.fromAHSL(
            (hsl1.alpha * (1 - amount)) + (hsl2.alpha * amount), h, s, l)
        .toColor();
  }

  // Whiten a color by mixing it with white by the provided amount (0-1), defaulting to 50%
  Color whiten([double amount = 0.5]) {
    return mix(Colors.white, amount);
  }

  //Tone down a color by mixing it with grey by the provided amount (0-1), defaulting to 50%
  Color tone([double amount = 0.5]) {
    const neutralGrey = Color.fromRGBO(128, 128, 128, 1);
    return mix(neutralGrey, amount);
  }

  // Blacken a color by mixing it with black by the provided amount (0-1), defaulting to 50%
  Color blacken([double amount = 0.5]) {
    return mix(Colors.black, amount);
  }

  // Check if a color is considered dark using the luminance
  bool get isDark {
    // Compute the color luminance
    final double luminance = 0.299 * red + 0.587 * green + 0.114 * blue;
    // Luminance ranges from 0 to 255, we take 128 as a mid-point
    return luminance < 128;
  }

  // Check if a color is considered bright
  bool get isBright {
    // Bright colors are not dark!
    return !isDark;
  }
}


lib/color_adjust.dart
import 'package:flutter/material.dart';

/// Extension on Color to add color manipulation methods
extension Adjust on Color {
  /// Darken a color by the provided amount (0-1), defaulting to 50%.
  Color darken([double amount = .5]) {
    assert(amount >= 0 && amount <= 1);
    final hsl = HSLColor.fromColor(this); // convert to HSL
    // darken by reducing lightness
    final hslDark = hsl.withLightness((hsl.lightness - amount).clamp(0.0, 1.0));
    return hslDark.toColor(); // convert back to Color
  }

  /// Lighten a color by the provided amount (0-1), defaulting to 50%.
  Color lighten([double amount = .5]) {
    assert(amount >= 0 && amount <= 1);
    final hsl = HSLColor.fromColor(this); // convert to HSL
    // lighten by increasing lightness
    final hslLight =
        hsl.withLightness((hsl.lightness + amount).clamp(0.0, 1.0));
    return hslLight.toColor(); // convert back to Color
  }

  /// Invert a color (producing its "negative").
  /// This will not preserve lightness. If you want to do that, use compliment
  /// to convert to HSL, invert the hue, and convert back to RGB.
  Color invert() {
    final r = (255 - red).clamp(0, 255);
    final g = (255 - green).clamp(0, 255);
    final b = (255 - blue).clamp(0, 255);
    return Color.fromARGB((opacity * 255).round(), r, g, b);
  }

  /// Get the complimentary (opposite on the color wheel) of a color.
  /// This is like inverting, however, it keeps the same lightness value.
  Color compliment() {
    final hsl = HSLColor.fromColor(this); // convert to HSL
    final double newHue =
        (hsl.hue + 180.0) % 360.0; // compute complimentary hue
    final newHsl =
        hsl.withHue(newHue); // create new color with complimentary hue
    return newHsl.toColor(); // convert back to Color
  }

  /// Desaturate a color by the provided amount (0-1), defaulting to 50%.
  Color desaturate([double amount = .5]) {
    assert(amount >= 0 && amount <= 1);
    final hsl = HSLColor.fromColor(this); // convert to HSL
    // desaturate by reducing saturation
    final hslDesaturated =
        hsl.withSaturation((hsl.saturation - amount).clamp(0.0, 1.0));
    return hslDesaturated.toColor(); // convert back to Color
  }

  /// Saturate a color by the provided amount (0-1), defaulting to 50%.
  Color saturate([double amount = .5]) {
    assert(amount >= 0 && amount <= 1);
    final hsl = HSLColor.fromColor(this); // convert to HSL
    // saturate by increasing saturation
    final hslSaturated =
        hsl.withSaturation((hsl.saturation + amount).clamp(0.0, 1.0));
    return hslSaturated.toColor(); // convert back to Color
  }

  /// Mix a color with another one by the provided amount (0-1), defaulting to 50%.
  /// This might result in a color that looks muddy, as the lightness is not preserved.
  /// If you want to preserve the lightness, use blend instead.
  Color mix(Color another, [double amount = 0.5]) {
    assert(amount >= 0 && amount <= 1);
    // mix by taking a weighted average
    final r = (red * (1 - amount) + another.red * amount).round().clamp(0, 255);
    final g =
        (green * (1 - amount) + another.green * amount).round().clamp(0, 255);
    final b =
        (blue * (1 - amount) + another.blue * amount).round().clamp(0, 255);
    return Color.fromARGB(alpha, r, g, b);
  }

  /// Mix a color with another one using HSL color space by the provided amount (0-1), defaulting to 50%.
  Color blend(Color another, [double amount = 0.5]) {
    assert(amount >= 0 && amount <= 1);
    final hsl1 = HSLColor.fromColor(this);
    final hsl2 = HSLColor.fromColor(another);
    // mix by taking a weighted average
    final h =
        ((hsl1.hue * (1 - amount)) + (hsl2.hue * amount)).clamp(0.0, 360.0);
    final s = ((hsl1.saturation * (1 - amount)) + (hsl2.saturation * amount))
        .clamp(0.0, 1.0);
    final l = ((hsl1.lightness * (1 - amount)) + (hsl2.lightness * amount))
        .clamp(0.0, 1.0);
    return HSLColor.fromAHSL(
            (hsl1.alpha * (1 - amount)) + (hsl2.alpha * amount), h, s, l)
        .toColor();
  }

  /// Whiten a color by mixing it with white by the provided amount (0-1), defaulting to 50%.
  Color whiten([double amount = 0.5]) {
    return mix(Colors.white, amount);
  }

  /// Tone down a color by mixing it with grey by the provided amount (0-1), defaulting to 50%.
  Color tone([double amount = 0.5]) {
    const neutralGrey = Color.fromRGBO(128, 128, 128, 1);
    return mix(neutralGrey, amount);
  }

  /// Blacken a color by mixing it with black by the provided amount (0-1), defaulting to 50%.
  Color blacken([double amount = 0.5]) {
    return mix(Colors.black, amount);
  }

  /// Check if a color is considered dark using the luminance.
  bool get isDark {
    // Compute the color luminance
    final double luminance = 0.299 * red + 0.587 * green + 0.114 * blue;
    // Luminance ranges from 0 to 255, we take 128 as a mid-point
    return luminance < 128;
  }

  /// Check if a color is considered bright.
  bool get isBright {
    // Bright colors are not dark!
    return !isDark;
  }
}


example/main.dart
import 'package:color_adjust/color_adjust.dart';
import 'package:flutter/material.dart';
import 'package:flutter_colorpicker/flutter_colorpicker.dart';

void main() {
  runApp(const MaterialApp(debugShowCheckedModeBanner: false, home: MyApp()));
}

class MyApp extends StatefulWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  MyAppState createState() => MyAppState();
}

class MyAppState extends State<MyApp> {
  Color baseColor = Colors.blue;
  Color mixColor = Colors.pink;
  late Color tempColor;
  TextStyle textLight = const TextStyle(color: Colors.white);
  TextStyle textDark = const TextStyle(color: Colors.black);

  void changeColor(Color color) {
    setState(() {
      tempColor = color;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text("Color Adjust Examples"),
      ),
      body: Column(
        children: [
          Wrap(
            runSpacing: 10,
            spacing: 10,
            runAlignment: WrapAlignment.center,
            children: <Widget>[
              ElevatedButton(
                style: ElevatedButton.styleFrom(backgroundColor: baseColor),
                child: Text(
                  'Select Base Color',
                  style: baseColor.isBright ? textDark : textLight,
                ),
                onPressed: () {
                  tempColor = baseColor;
                  showDialog(
                    context: context,
                    builder: (BuildContext context) {
                      return AlertDialog(
                        title: const Text('Pick a color'),
                        content: SingleChildScrollView(
                          child: ColorPicker(
                            pickerColor: tempColor,
                            onColorChanged: changeColor,
                            colorPickerWidth: 300.0,
                            pickerAreaHeightPercent: 0.7,
                            enableAlpha: true,
                            displayThumbColor: true,
                            paletteType: PaletteType.hsv,
                            pickerAreaBorderRadius: const BorderRadius.only(
                              topLeft: Radius.circular(2.0),
                              topRight: Radius.circular(2.0),
                            ),
                          ),
                        ),
                        actions: <Widget>[
                          TextButton(
                            child: const Text('Got it'),
                            onPressed: () {
                              setState(() {
                                baseColor = tempColor;
                              });
                              Navigator.of(context).pop();
                            },
                          ),
                        ],
                      );
                    },
                  );
                },
              ),
              ElevatedButton(
                style: ElevatedButton.styleFrom(backgroundColor: mixColor),
                child: Text(
                  'Select Mix Color',
                  style: mixColor.isBright ? textDark : textLight,
                ),
                onPressed: () {
                  tempColor = mixColor;
                  showDialog(
                    context: context,
                    builder: (BuildContext context) {
                      return AlertDialog(
                        title: const Text('Pick a color'),
                        content: SingleChildScrollView(
                          child: ColorPicker(
                            pickerColor: tempColor,
                            onColorChanged: changeColor,
                            colorPickerWidth: 300.0,
                            pickerAreaHeightPercent: 0.7,
                            enableAlpha: true,
                            displayThumbColor: true,
                            paletteType: PaletteType.hsv,
                            pickerAreaBorderRadius: const BorderRadius.only(
                              topLeft: Radius.circular(2.0),
                              topRight: Radius.circular(2.0),
                            ),
                          ),
                        ),
                        actions: <Widget>[
                          TextButton(
                            child: const Text('Got it'),
                            onPressed: () {
                              setState(() {
                                mixColor = tempColor;
                              });
                              Navigator.of(context).pop();
                            },
                          ),
                        ],
                      );
                    },
                  );
                },
              ),
            ],
          ),
          Expanded(
            child: getList(),
          ),
        ],
      ),
    );
  }

  Widget getList() {
    return ListView(
      children: <Widget>[
        Builder(builder: (context) {
          return Row(
            children: [
              Expanded(
                child: Container(
                    color: baseColor.mix(mixColor, 0.5),
                    alignment: Alignment.center,
                    height: 50,
                    child: Text(
                      "Mix Default",
                      style: baseColor.mix(mixColor, 0.5).isDark
                          ? textLight
                          : textDark,
                    )),
              ),
              Expanded(
                child: Container(
                    color: baseColor.mix(mixColor, 0.25),
                    alignment: Alignment.center,
                    height: 50,
                    child: Text(
                      "Mix 25%",
                      style: baseColor.mix(mixColor, 0.25).isDark
                          ? textLight
                          : textDark,
                    )),
              ),
              Expanded(
                child: Container(
                    color: baseColor.mix(mixColor, 0.75),
                    alignment: Alignment.center,
                    height: 50,
                    child: Text(
                      "Mix 75%",
                      style: baseColor.mix(mixColor, 0.75).isDark
                          ? textLight
                          : textDark,
                    )),
              ),
            ],
          );
        }),
        Builder(builder: (context) {
          return Row(
            children: [
              Expanded(
                child: Container(
                    color: baseColor.blend(mixColor),
                    alignment: Alignment.center,
                    height: 50,
                    child: Text(
                      "Blend Default",
                      style: baseColor.blend(mixColor).isDark
                          ? textLight
                          : textDark,
                    )),
              ),
              Expanded(
                child: Container(
                    color: baseColor.blend(mixColor, .25),
                    alignment: Alignment.center,
                    height: 50,
                    child: Text(
                      "Blend 25%",
                      style: baseColor.blend(mixColor, .25).isDark
                          ? textLight
                          : textDark,
                    )),
              ),
              Expanded(
                child: Container(
                    color: baseColor.blend(mixColor, .75),
                    alignment: Alignment.center,
                    height: 50,
                    child: Text(
                      "Blend 75%",
                      style: baseColor.blend(mixColor, .75).isDark
                          ? textLight
                          : textDark,
                    )),
              ),
            ],
          );
        }),
        Builder(builder: (context) {
          final Color darkened = baseColor.darken(0.2);
          return Container(
              color: darkened,
              alignment: Alignment.center,
              height: 50,
              child: Text(
                "Darkened",
                style: darkened.isDark ? textLight : textDark,
              ));
        }),
        Builder(builder: (context) {
          final Color lightened = baseColor.lighten(0.2);
          return Container(
              color: lightened,
              alignment: Alignment.center,
              height: 50,
              child: Text(
                "Lightened",
                style: lightened.isDark ? textLight : textDark,
              ));
        }),
        Builder(builder: (context) {
          final Color inverted = baseColor.invert();
          return Container(
              color: inverted,
              alignment: Alignment.center,
              height: 50,
              child: Text(
                "Inverted",
                style: inverted.isDark ? textLight : textDark,
              ));
        }),
        Builder(builder: (context) {
          final Color desaturated = baseColor.desaturate(.5);
          return Container(
              color: desaturated,
              alignment: Alignment.center,
              height: 50,
              child: Text(
                "Desaturated",
                style: desaturated.isDark ? textLight : textDark,
              ));
        }),
        Builder(builder: (context) {
          final Color saturated = baseColor.saturate(.5);
          return Container(
              color: saturated,
              alignment: Alignment.center,
              height: 50,
              child: Text(
                "Saturated",
                style: saturated.isDark ? textLight : textDark,
              ));
        }),
        Builder(builder: (context) {
          final Color complimented = baseColor.compliment();
          return Container(
              color: complimented,
              alignment: Alignment.center,
              height: 50,
              child: Text(
                "Complimented",
                style: complimented.isDark ? textLight : textDark,
              ));
        }),
        Builder(builder: (context) {
          final Color whitened = baseColor.whiten(0.5);
          return Container(
              color: whitened,
              alignment: Alignment.center,
              height: 50,
              child: Text(
                "Whitened",
                style: whitened.isDark ? textLight : textDark,
              ));
        }),
        Builder(builder: (context) {
          final Color toned = baseColor.tone(0.5);
          return Container(
              color: toned,
              alignment: Alignment.center,
              height: 50,
              child: Text(
                "Toned",
                style: toned.isDark ? textLight : textDark,
              ));
        }),
        Builder(builder: (context) {
          final Color blackened = baseColor.blacken(0.5);
          return Container(
              color: blackened,
              alignment: Alignment.center,
              height: 50,
              child: Text(
                "Blackened",
                style: blackened.isDark ? textLight : textDark,
              ));
        }),
      ],
    );
  }
}
